source/extensions/filters/http/ext_proc/client_impl.cc

```diff
@@ -14,21 +14,25 @@ ExternalProcessorClientImpl::ExternalProcessorClientImpl(Grpc::AsyncClientManage
 ExternalProcessorStreamPtr
 ExternalProcessorClientImpl::start(ExternalProcessorCallbacks& callbacks,
                                    const Grpc::GrpcServiceConfigWithHashKey& config_with_hash_key,
-                                   const Http::AsyncClient::StreamOptions& options) {
+                                   const Http::AsyncClient::StreamOptions& options,
+                                   Http::StreamDecoderFilterCallbacks* decoder_filter_callbacks) {
   auto client_or_error =
       client_manager_.getOrCreateRawAsyncClientWithHashKey(config_with_hash_key, scope_, true);
   THROW_IF_STATUS_NOT_OK(client_or_error, throw);
   Grpc::AsyncClient<ProcessingRequest, ProcessingResponse> grpcClient(client_or_error.value());
-  return ExternalProcessorStreamImpl::create(std::move(grpcClient), callbacks, options);
+  return ExternalProcessorStreamImpl::create(std::move(grpcClient), callbacks, options,
+                                             decoder_filter_callbacks);
 }
 
 ExternalProcessorStreamPtr ExternalProcessorStreamImpl::create(
     Grpc::AsyncClient<ProcessingRequest, ProcessingResponse>&& client,
-    ExternalProcessorCallbacks& callbacks, const Http::AsyncClient::StreamOptions& options) {
+    ExternalProcessorCallbacks& callbacks, const Http::AsyncClient::StreamOptions& options,
+    Http::StreamDecoderFilterCallbacks* decoder_filter_callbacks) {
   auto stream =
       std::unique_ptr<ExternalProcessorStreamImpl>(new ExternalProcessorStreamImpl(callbacks));
 
   if (stream->startStream(std::move(client), options)) {
+    stream->stream_->setWatermarkCallbacks(*decoder_filter_callbacks);
```

Comment:
optional but I tend to runtime guard all things watermarks since they're complicated 
this one seems pretty straight-forward so your call

Reply:
one flag SGTM but let's add it now so the new code is protected as we land it?  I think we'll need to latch a bool to make sure we consistently register and clear callbacks

Reply:
Sure SG! Done.